File: android/app/src/main/java/com/ayuvo/health/insights/InsightsConfig.kt
package com.ayuvo.health.insights

import com.ayuvo.health.medications.logic.MedicationJson
import com.ayuvo.health.medications.logic.MedicationJson.arr
import com.ayuvo.health.medications.logic.MedicationJson.double
import com.ayuvo.health.medications.logic.MedicationJson.int
import com.ayuvo.health.medications.logic.MedicationJson.objOrNull
import com.ayuvo.health.medications.logic.MedicationJson.str
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull

/** One `[x, y]` point of a piecewise-linear table (reference tables, dose-response curves). */
typealias Points = List<Pair<Double, Double>>

/**
 * `assets/insights/insights_config.json`, a byte copy of `shared/insights/insights_config.json`
 * (docs/insights.md). Every weight, window, band, template and methodology text the engines and the
 * info sheets use comes from here; `scripts/insights_reference.py` is the normative algorithm.
 */
class InsightsConfig(val root: JsonObject) {

    data class Metric(
        val id: String,
        val direction: String,
        val androidType: String?,
        val label: String,
        val unit: String,
        val windowDays: Int,
        val minPoints: Int,
        val sdFloor: Double,
        val recent: String,
        val highConfidenceCoverage: Double,
        val highConfidenceMinN: Int,
        val trendDays: Int,
        val trendMinPoints: Int,
        val trendStablePctPerWeek: Double
    )

    data class RecoveryComponentSpec(
        val id: String,
        val metric: String,
        val mode: String,
        val weight: Double,
        val toleranceZ: Double,
        val durationShare: Double,
        val consistencyShare: Double
    )

    data class Band(val id: String, val label: String, val min: Int, val recommendation: String)

    data class Recovery(
        val bands: List<Band>,
        val components: List<RecoveryComponentSpec>,
        val consistencyMinNights: Int,
        val consistencyWindowDays: Int,
        val consistencyZeroAtMin: Double,
        val contributors: Map<String, String>,
        val heartComponents: List<String>,
        val loadHigh: Int,
        val loadVeryHigh: Int,
        val loadVeryHighRatio: Double,
        val subscoreCenter: Double,
        val subscoreSlope: Double,
        val source: String
    )

    data class TrainingLoad(
        val defaultIntensity: Double,
        val effortDivisor: Double,
        val highAbove: Double,
        val lightBelow: Double,
        val intenseMinIntensity: Double,
        val intenseMinMinutesWithoutEffort: Double,
        val intensityMin: Double,
        val intensityMax: Double,
        val labels: Map<String, String>,
        val lateHour: Int,
        val meanWindowDays: Int,
        val source: String
    )

    data class Marker(
        val id: String,
        val label: String,
        val method: String,
        val weight: Double,
        val minDays: Int,
        val capYears: Double,
        val series: String?,
        val unit: String,
        val source: String,
        /** `age_norm`: sex → table, or HRV kind → sex → table when [tablesByKind]. */
        val tables: Map<String, Points>,
        val tablesByKind: Map<String, Map<String, Points>>,
        val points: Points,
        val durationPoints: Points,
        val regularityPoints: Points,
        val minutesPoints: Points,
        val activeSharePoints: Points,
        val weeks: Int,
        val bmiPoints: Points,
        val bodyFatPoints: Map<String, Points>
    )

    data class HealthAge(
        val ageMin: Double,
        val ageMax: Double,
        val collectingDays: Int,
        val highMinMarkers: Int,
        val highRequires: String,
        val mediumMinMarkers: Int,
        val coreMarkers: List<String>,
        val daysPerYear: Double,
        val markers: List<Marker>,
        val minActualAge: Double,
        val minMarkers: Int,
        val paceDecliningAbove: Double,
        val paceImprovingBelow: Double,
        val paceMinPoints: Int,
        val paceWeeks: Int,
        val totalCapYears: Double,
        val windowDays: Int,
        val source: String
    )

    data class Area(val id: String, val label: String, val weight: Double)

    data class Rule(val id: String, val category: String, val template: String, val raw: JsonObject) {
        fun num(key: String): Double = raw.double(key) ?: error("rule $id has no $key")
    }

    data class ReduceNutrient(val id: String, val goalKey: String, val label: String, val unit: String)

    data class DailyReview(
        val areas: List<Area>,
        val averageMinDays: Int,
        val averageWindowDays: Int,
        val maxItemsPerCategory: Int,
        val notLoggedTemplate: String,
        val calorieTolerancePct: Double,
        val calorieZeroAtPct: Double,
        val macroTolerancePct: Double,
        val macroZeroAtPct: Double,
        val proteinMinShare: Double,
        val reduceNutrients: List<ReduceNutrient>,
        val rules: List<Rule>,
        val sleepTargetMin: Double,
        val trainingRestScore: Double
    )

    data class PatternPair(
        val id: String,
        val exposure: String,
        val outcome: String,
        val lagDays: Int,
        val decimals: Int,
        val unit: String,
        val moreWord: String,
        val lessWord: String,
        val template: String,
        val reviewCategory: String?
    )

    data class Patterns(
        val minAbsD: Double,
        val minAbsT: Double,
        val minGroup: Int,
        val pairs: List<PatternPair>,
        val partialTodayOutcomes: Set<String>,
        val shortSleepMarginMin: Double,
        val windowDays: Int,
        val source: String
    )

    data class Section(val heading: String, val body: String)
    data class Methodology(val title: String, val sections: List<Section>)

    data class Ai(
        val allowedNumbers: List<Double>,
        val blockedTerms: List<String>,
        val bulletMax: Int,
        val bulletsMax: Int,
        val headlineMax: Int,
        val maxOutputTokensCloud: Int,
        val maxOutputTokensLocal: Int,
        val statusCloud: String,
        val statusLocal: String,
        val tasks: Map<String, String>
    )

    val metrics: Map<String, Metric> = obj(root, "metrics").entries.associate { (id, v) -> id to metric(id, v as JsonObject) }
    val recovery: Recovery = recovery(obj(root, "recovery"))
    val trainingLoad: TrainingLoad = trainingLoad(obj(root, "training_load"))
    val healthAge: HealthAge = healthAge(obj(root, "health_age"))
    val dailyReview: DailyReview = dailyReview(obj(root, "daily_review"))
    val patterns: Patterns = patterns(obj(root, "patterns"))
    val minNightMinutes: Double = obj(root, "sleep").double("min_night_minutes")!!
    val disclaimers: Map<String, String> = obj(root, "disclaimers").entries.associate { it.key to (it.value as JsonPrimitive).content }
    val methodology: Map<String, Methodology> = obj(root, "methodology").entries.associate { (id, v) ->
        val o = v as JsonObject
        id to Methodology(o.str("title")!!, o.arr("sections")!!.map { s -> (s as JsonObject).let { Section(it.str("heading")!!, it.str("body")!!) } })
    }
    val ai: Ai = ai(obj(root, "ai"))

    fun metric(id: String): Metric = metrics[id] ?: error("unknown metric $id")

    companion object {
        const val ASSET_PATH = "insights/insights_config.json"
        const val PROMPT_ASSET_PATH = "insights/ai_explain.md"

        @Volatile
        var active: InsightsConfig? = null

        fun parse(text: String): InsightsConfig = InsightsConfig(MedicationJson.json.parseToJsonElement(text) as JsonObject)

        private fun obj(o: JsonObject, key: String): JsonObject = o.objOrNull(key) ?: error("config has no $key")

        private fun points(e: JsonElement?): Points = (e as? JsonArray).orEmpty().map { p ->
            val a = p as JsonArray
            (a[0] as JsonPrimitive).doubleOrNull!! to (a[1] as JsonPrimitive).doubleOrNull!!
        }

        private fun sexPoints(e: JsonElement?): Map<String, Points> =
            (e as? JsonObject)?.entries?.associate { it.key to points(it.value) }.orEmpty()

        private fun strings(e: JsonElement?): List<String> = MedicationJson.strings(e)

        private fun metric(id: String, o: JsonObject) = Metric(
            id = id,
            direction = o.str("direction")!!,
            androidType = o.objOrNull("health_type")?.str("android"),
            label = o.str("label")!!,
            unit = o.str("unit")!!,
            windowDays = o.int("window_days")!!,
            minPoints = o.int("min_points")!!,
            sdFloor = o.double("sd_floor")!!,
            recent = o.str("recent")!!,
            highConfidenceCoverage = o.double("high_confidence_coverage")!!,
            highConfidenceMinN = o.int("high_confidence_min_n")!!,
            trendDays = o.int("trend_days")!!,
            trendMinPoints = o.int("trend_min_points")!!,
            trendStablePctPerWeek = o.double("trend_stable_pct_per_week")!!
        )

        private fun recovery(o: JsonObject): Recovery {
            val lm = obj(o, "load_modifier")
            return Recovery(
                bands = o.arr("bands")!!.map { (it as JsonObject).let { b -> Band(b.str("id")!!, b.str("label")!!, b.int("min")!!, b.str("recommendation")!!) } },
                components = o.arr("components")!!.map {
                    val c = it as JsonObject
                    RecoveryComponentSpec(
                        id = c.str("id")!!, metric = c.str("metric")!!, mode = c.str("mode")!!, weight = c.double("weight")!!,
                        toleranceZ = c.double("tolerance_z") ?: 0.0, durationShare = c.double("duration_share") ?: 1.0,
                        consistencyShare = c.double("consistency_share") ?: 0.0
                    )
                },
                consistencyMinNights = o.int("consistency_min_nights")!!,
                consistencyWindowDays = o.int("consistency_window_days")!!,
                consistencyZeroAtMin = o.double("consistency_zero_at_min")!!,
                contributors = obj(o, "contributors").entries.associate { it.key to (it.value as JsonPrimitive).content },
                heartComponents = strings(o["heart_components"]),
                loadHigh = lm.int("high")!!,
                loadVeryHigh = lm.int("very_high")!!,
                loadVeryHighRatio = lm.double("very_high_ratio")!!,
                subscoreCenter = o.double("subscore_center")!!,
                subscoreSlope = o.double("subscore_slope")!!,
                source = o.str("source").orEmpty()
            )
        }

        private fun trainingLoad(o: JsonObject) = TrainingLoad(
            defaultIntensity = o.double("default_intensity")!!,
            effortDivisor = o.double("effort_divisor")!!,
            highAbove = o.double("high_above")!!,
            lightBelow = o.double("light_below")!!,
            intenseMinIntensity = o.double("intense_min_intensity")!!,
            intenseMinMinutesWithoutEffort = o.double("intense_min_minutes_without_effort")!!,
            intensityMin = o.double("intensity_min")!!,
            intensityMax = o.double("intensity_max")!!,
            labels = obj(o, "labels").entries.associate { it.key to (it.value as JsonPrimitive).content },
            lateHour = o.int("late_hour")!!,
            meanWindowDays = o.int("mean_window_days")!!,
            source = o.str("source").orEmpty()
        )

        private fun marker(o: JsonObject): Marker {
            val byKind = (o["tables_by_kind"] as? JsonPrimitive)?.booleanOrNull == true
            val tables = o.objOrNull("tables")
            return Marker(
                id = o.str("id")!!,
                label = o.str("label")!!,
                method = o.str("method")!!,
                weight = o.double("weight")!!,
                minDays = o.int("min_days")!!,
                capYears = o.double("cap_years")!!,
                series = o.str("series"),
                unit = o.str("unit").orEmpty(),
                source = o.str("source").orEmpty(),
                tables = if (byKind) emptyMap() else sexPoints(tables),
                tablesByKind = if (byKind) tables!!.entries.associate { it.key to sexPoints(it.value) } else emptyMap(),
                points = points(o["points"]),
                durationPoints = points(o["duration_points"]),
                regularityPoints = points(o["regularity_points"]),
                minutesPoints = points(o["minutes_points"]),
                activeSharePoints = points(o["active_share_points"]),
                weeks = o.int("weeks") ?: 0,
                bmiPoints = points(o["bmi_points"]),
                bodyFatPoints = sexPoints(o["body_fat_points"])
            )
        }

        private fun healthAge(o: JsonObject): HealthAge {
            val c = obj(o, "confidence")
            val p = obj(o, "pace")
            return HealthAge(
                ageMin = o.double("age_min")!!,
                ageMax = o.double("age_max")!!,
                collectingDays = o.int("collecting_days")!!,
                highMinMarkers = c.int("high_min_markers")!!,
                highRequires = c.str("high_requires")!!,
                mediumMinMarkers = c.int("medium_min_markers")!!,
                coreMarkers = strings(o["core_markers"]),
                daysPerYear = o.double("days_per_year")!!,
                markers = o.arr("markers")!!.map { marker(it as JsonObject) },
                minActualAge = o.double("min_actual_age")!!,
                minMarkers = o.int("min_markers")!!,
                paceDecliningAbove = p.double("declining_above")!!,
                paceImprovingBelow = p.double("improving_below")!!,
                paceMinPoints = p.int("min_points")!!,
                paceWeeks = p.int("weeks")!!,
                totalCapYears = o.double("total_cap_years")!!,
                windowDays = o.int("window_days")!!,
                source = o.str("source").orEmpty()
            )
        }

        private fun dailyReview(o: JsonObject): DailyReview {
            val n = obj(o, "nutrition")
            return DailyReview(
                areas = o.arr("areas")!!.map { (it as JsonObject).let { a -> Area(a.str("id")!!, a.str("label")!!, a.double("weight")!!) } },
                averageMinDays = o.int("average_min_days")!!,
                averageWindowDays = o.int("average_window_days")!!,
                maxItemsPerCategory = o.int("max_items_per_category")!!,
                notLoggedTemplate = o.str("not_logged_template")!!,
                calorieTolerancePct = n.double("calorie_tolerance_pct")!!,
                calorieZeroAtPct = n.double("calorie_zero_at_pct")!!,
                macroTolerancePct = n.double("macro_tolerance_pct")!!,
                macroZeroAtPct = n.double("macro_zero_at_pct")!!,
                proteinMinShare = n.double("protein_min_share")!!,
                reduceNutrients = o.arr("reduce_nutrients")!!.map {
                    (it as JsonObject).let { r -> ReduceNutrient(r.str("id")!!, r.str("goal_key")!!, r.str("label")!!, r.str("unit")!!) }
                },
                rules = o.arr("rules")!!.map { (it as JsonObject).let { r -> Rule(r.str("id")!!, r.str("category")!!, r.str("template")!!, r) } },
                sleepTargetMin = o.double("sleep_target_min")!!,
                trainingRestScore = o.double("training_rest_score")!!
            )
        }

        private fun patterns(o: JsonObject) = Patterns(
            minAbsD = o.double("min_abs_d")!!,
            minAbsT = o.double("min_abs_t")!!,
            minGroup = o.int("min_group")!!,
            pairs = o.arr("pairs")!!.map {
                val p = it as JsonObject
                PatternPair(
                    id = p.str("id")!!, exposure = p.str("exposure")!!, outcome = p.str("outcome")!!, lagDays = p.int("lag_days")!!,
                    decimals = p.int("decimals")!!, unit = p.str("unit")!!, moreWord = p.str("more_word")!!, lessWord = p.str("less_word")!!,
                    template = p.str("template")!!, reviewCategory = p.str("review_category")
                )
            },
            partialTodayOutcomes = strings(o["partial_today_outcomes"]).toSet(),
            shortSleepMarginMin = o.double("short_sleep_margin_min")!!,
            windowDays = o.int("window_days")!!,
            source = o.str("source").orEmpty()
        )

        private fun ai(o: JsonObject): Ai {
            val tokens = obj(o, "max_output_tokens")
            val labels = obj(o, "status_labels")
            return Ai(
                allowedNumbers = (o["allowed_numbers"] as? JsonArray).orEmpty().mapNotNull { (it as? JsonPrimitive)?.doubleOrNull },
                blockedTerms = strings(o["blocked_terms"]),
                bulletMax = o.int("bullet_max")!!,
                bulletsMax = o.int("bullets_max")!!,
                headlineMax = o.int("headline_max")!!,
                maxOutputTokensCloud = tokens.int("cloud")!!,
                maxOutputTokensLocal = tokens.int("local")!!,
                statusCloud = labels.str("cloud")!!,
                statusLocal = labels.str("local")!!,
                tasks = obj(o, "tasks").entries.associate { it.key to (it.value as JsonPrimitive).content }
            )
        }
    }
}

File: ios/calorietracker/Insights/InsightsConfig.swift
import Foundation

/// Typed view of `shared/insights/insights_config.json` (bundled as `Insights/Resources/insights_config.json`,
/// byte-identical to the shared file). Every constant the engines use comes from here.
nonisolated struct InsightsConfig: Decodable, Sendable {
    let configVersion: Int
    let ai: AI
    let dailyReview: DailyReview
    let disclaimers: [String: String]
    let healthAge: HealthAge
    let methodology: [String: Methodology]
    let metrics: [String: Metric]
    let patterns: Patterns
    let recovery: Recovery
    let sleep: Sleep
    let trainingLoad: TrainingLoad

    enum CodingKeys: String, CodingKey {
        case configVersion = "config_version", ai, dailyReview = "daily_review", disclaimers,
             healthAge = "health_age", methodology, metrics, patterns, recovery, sleep, trainingLoad = "training_load"
    }

    // MARK: Sections

    struct AI: Decodable, Sendable {
        let allowedNumbers: [Double]
        let blockedTerms: [String]
        let bulletMax: Int
        let bulletsMax: Int
        let headlineMax: Int
        let maxOutputTokens: [String: Int]
        let statusLabels: [String: String]
        let tasks: [String: String]

        enum CodingKeys: String, CodingKey {
            case allowedNumbers = "allowed_numbers", blockedTerms = "blocked_terms", bulletMax = "bullet_max",
                 bulletsMax = "bullets_max", headlineMax = "headline_max", maxOutputTokens = "max_output_tokens",
                 statusLabels = "status_labels", tasks
        }
    }

    struct Metric: Decodable, Sendable {
        let direction: String
        let healthType: [String: String?]
        let highConfidenceCoverage: Double
        let highConfidenceMinN: Int
        let label: String
        let minPoints: Int
        let recent: String
        let sdFloor: Double
        let trendDays: Int
        let trendMinPoints: Int
        let trendStablePctPerWeek: Double
        let unit: String
        let windowDays: Int

        enum CodingKeys: String, CodingKey {
            case direction, healthType = "health_type", highConfidenceCoverage = "high_confidence_coverage",
                 highConfidenceMinN = "high_confidence_min_n", label, minPoints = "min_points", recent,
                 sdFloor = "sd_floor", trendDays = "trend_days", trendMinPoints = "trend_min_points",
                 trendStablePctPerWeek = "trend_stable_pct_per_week", unit, windowDays = "window_days"
        }

        var iosHealthType: String? { healthType["ios"] ?? nil }
    }

    struct Sleep: Decodable, Sendable {
        let minNightMinutes: Double
        enum CodingKeys: String, CodingKey { case minNightMinutes = "min_night_minutes" }
    }

    struct TrainingLoad: Decodable, Sendable {
        let defaultIntensity: Double
        let effortDivisor: Double
        let highAbove: Double
        let intenseMinIntensity: Double
        let intenseMinMinutesWithoutEffort: Double
        let intensityMax: Double
        let intensityMin: Double
        let labels: [String: String]
        let lateHour: Int
        let lightBelow: Double
        let meanWindowDays: Int
        let source: String

        enum CodingKeys: String, CodingKey {
            case defaultIntensity = "default_intensity", effortDivisor = "effort_divisor", highAbove = "high_above",
                 intenseMinIntensity = "intense_min_intensity",
                 intenseMinMinutesWithoutEffort = "intense_min_minutes_without_effort", intensityMax = "intensity_max",
                 intensityMin = "intensity_min", labels, lateHour = "late_hour", lightBelow = "light_below",
                 meanWindowDays = "mean_window_days", source
        }
    }

    struct Recovery: Decodable, Sendable {
        struct Band: Decodable, Sendable {
            let id: String
            let label: String
            let min: Int
            let recommendation: String
        }

        struct Component: Decodable, Sendable {
            let id: String
            let metric: String
            let mode: String
            let weight: Double
            let toleranceZ: Double?
            let durationShare: Double?
            let consistencyShare: Double?

            enum CodingKeys: String, CodingKey {
                case id, metric, mode, weight, toleranceZ = "tolerance_z", durationShare = "duration_share",
                     consistencyShare = "consistency_share"
            }
        }

        struct LoadModifier: Decodable, Sendable {
            let high: Int
            let veryHigh: Int
            let veryHighRatio: Double
            enum CodingKeys: String, CodingKey { case high, veryHigh = "very_high", veryHighRatio = "very_high_ratio" }
        }

        let bands: [Band]
        let components: [Component]
        let consistencyMinNights: Int
        let consistencyWindowDays: Int
        let consistencyZeroAtMin: Double
        let contributors: [String: String]
        let heartComponents: [String]
        let loadModifier: LoadModifier
        let source: String
        let subscoreCenter: Double
        let subscoreSlope: Double

        enum CodingKeys: String, CodingKey {
            case bands, components, consistencyMinNights = "consistency_min_nights",
                 consistencyWindowDays = "consistency_window_days", consistencyZeroAtMin = "consistency_zero_at_min",
                 contributors, heartComponents = "heart_components", loadModifier = "load_modifier", source,
                 subscoreCenter = "subscore_center", subscoreSlope = "subscore_slope"
        }
    }

    struct HealthAge: Decodable, Sendable {
        struct Confidence: Decodable, Sendable {
            let highMinMarkers: Int
            let highRequires: String
            let mediumMinMarkers: Int
            enum CodingKeys: String, CodingKey {
                case highMinMarkers = "high_min_markers", highRequires = "high_requires",
                     mediumMinMarkers = "medium_min_markers"
            }
        }

        struct Pace: Decodable, Sendable {
            let decliningAbove: Double
            let improvingBelow: Double
            let minPoints: Int
            let weeks: Int
            enum CodingKeys: String, CodingKey {
                case decliningAbove = "declining_above", improvingBelow = "improving_below", minPoints = "min_points", weeks
            }
        }

        struct Marker: Decodable, Sendable {
            let id: String
            let label: String
            let method: String
            let minDays: Int
            let capYears: Double
            let weight: Double
            let unit: String
            let source: String
            let series: String?
            /// dose_response
            let points: [[Double]]?
            /// age_norm, not by kind: sex → [[age, value]]
            let tables: [String: [[Double]]]?
            /// age_norm by HRV kind: kind → sex → [[age, value]]
            let tablesByKind: [String: [String: [[Double]]]]?
            /// sleep
            let durationPoints: [[Double]]?
            let regularityPoints: [[Double]]?
            /// workouts
            let weeks: Int?
            let minutesPoints: [[Double]]?
            let activeSharePoints: [[Double]]?
            /// body composition
            let bmiPoints: [[Double]]?
            let bodyFatPoints: [String: [[Double]]]?

            enum CodingKeys: String, CodingKey {
                case id, label, method, minDays = "min_days", capYears = "cap_years", weight, unit, source, series,
                     points, tables, tablesByKind = "tables_by_kind", durationPoints = "duration_points",
                     regularityPoints = "regularity_points", weeks, minutesPoints = "minutes_points",
                     activeSharePoints = "active_share_points", bmiPoints = "bmi_points",
                     bodyFatPoints = "body_fat_points"
            }

            init(from decoder: Decoder) throws {
                let c = try decoder.container(keyedBy: CodingKeys.self)
                id = try c.decode(String.self, forKey: .id)
                label = try c.decode(String.self, forKey: .label)
                method = try c.decode(String.self, forKey: .method)
                minDays = try c.decode(Int.self, forKey: .minDays)
                capYears = try c.decode(Double.self, forKey: .capYears)
                weight = try c.decode(Double.self, forKey: .weight)
                unit = try c.decode(String.self, forKey: .unit)
                source = try c.decode(String.self, forKey: .source)
                series = try c.decodeIfPresent(String.self, forKey: .series)
                points = try c.decodeIfPresent([[Double]].self, forKey: .points)
                let byKind = try c.decodeIfPresent(Bool.self, forKey: .tablesByKind) ?? false
                if byKind {
                    tables = nil
                    tablesByKind = try c.decodeIfPresent([String: [String: [[Double]]]].self, forKey: .tables)
                } else {
                    tables = try c.decodeIfPresent([String: [[Double]]].self, forKey: .tables)
                    tablesByKind = nil
                }
                durationPoints = try c.decodeIfPresent([[Double]].self, forKey: .durationPoints)
                regularityPoints = try c.decodeIfPresent([[Double]].self, forKey: .regularityPoints)
                weeks = try c.decodeIfPresent(Int.self, forKey: .weeks)
                minutesPoints = try c.decodeIfPresent([[Double]].self, forKey: .minutesPoints)
                activeSharePoints = try c.decodeIfPresent([[Double]].self, forKey: .activeSharePoints)
                bmiPoints = try c.decodeIfPresent([[Double]].self, forKey: .bmiPoints)
                bodyFatPoints = try c.decodeIfPresent([String: [[Double]]].self, forKey: .bodyFatPoints)
            }
        }

        let ageMax: Double
        let ageMin: Double
        let collectingDays: Int
        let confidence: Confidence
        let coreMarkers: [String]
        let daysPerYear: Double
        let markers: [Marker]
        let minActualAge: Double
        let minMarkers: Int
        let pace: Pace
        let source: String
        let totalCapYears: Double
        let windowDays: Int

        enum CodingKeys: String, CodingKey {
            case ageMax = "age_max", ageMin = "age_min", collectingDays = "collecting_days", confidence,
                 coreMarkers = "core_markers", daysPerYear = "days_per_year", markers, minActualAge = "min_actual_age",
                 minMarkers = "min_markers", pace, source, totalCapYears = "total_cap_years", windowDays = "window_days"
        }
    }

    struct DailyReview: Decodable, Sendable {
        struct Area: Decodable, Sendable {
            let id: String
            let label: String
            let weight: Double
        }

        struct Nutrition: Decodable, Sendable {
            let calorieTolerancePct: Double
            let calorieZeroAtPct: Double
            let macroTolerancePct: Double
            let macroZeroAtPct: Double
            let proteinMinShare: Double
            enum CodingKeys: String, CodingKey {
                case calorieTolerancePct = "calorie_tolerance_pct", calorieZeroAtPct = "calorie_zero_at_pct",
                     macroTolerancePct = "macro_tolerance_pct", macroZeroAtPct = "macro_zero_at_pct",
                     proteinMinShare = "protein_min_share"
            }
        }

        struct ReduceNutrient: Decodable, Sendable {
            let goalKey: String
            let id: String
            let label: String
            let unit: String
            enum CodingKeys: String, CodingKey { case goalKey = "goal_key", id, label, unit }
        }

        struct Rule: Decodable, Sendable {
            let id: String
            let category: String
            let template: String
            let params: [String]
            let minShare: Double?
            let belowShare: Double?
            let minMinutes: Double?
            let belowMinutes: Double?
            let marginMinutes: Double?
            let stepsPerMinute: Double?
            let walkRoundMin: Int?
            let maxMinutes: Int?
            let roundMin: Int?

            enum CodingKeys: String, CodingKey {
                case id, category, template, params, minShare = "min_share", belowShare = "below_share",
                     minMinutes = "min_minutes", belowMinutes = "below_minutes", marginMinutes = "margin_minutes",
                     stepsPerMinute = "steps_per_minute", walkRoundMin = "walk_round_min", maxMinutes = "max_minutes",
                     roundMin = "round_min"
            }
        }

        let areas: [Area]
        let averageMinDays: Int
        let averageWindowDays: Int
        let maxItemsPerCategory: Int
        let notLoggedTemplate: String
        let nutrition: Nutrition
        let reduceNutrients: [ReduceNutrient]
        let rules: [Rule]
        let sleepTargetMin: Double
        let trainingRestScore: Double

        enum CodingKeys: String, CodingKey {
            case areas, averageMinDays = "average_min_days", averageWindowDays = "average_window_days",
                 maxItemsPerCategory = "max_items_per_category", notLoggedTemplate = "not_logged_template", nutrition,
                 reduceNutrients = "reduce_nutrients", rules, sleepTargetMin = "sleep_target_min",
                 trainingRestScore = "training_rest_score"
        }
    }

    struct Patterns: Decodable, Sendable {
        struct Pair: Decodable, Sendable {
            let id: String
            let exposure: String
            let outcome: String
            let lagDays: Int
            let decimals: Int
            let unit: String
            let moreWord: String
            let lessWord: String
            let template: String
            let reviewCategory: String?

            enum CodingKeys: String, CodingKey {
                case id, exposure, outcome, lagDays = "lag_days", decimals, unit, moreWord = "more_word",
                     lessWord = "less_word", template, reviewCategory = "review_category"
            }
        }

        let minAbsD: Double
        let minAbsT: Double
        let minGroup: Int
        let pairs: [Pair]
        let partialTodayOutcomes: [String]
        let shortSleepMarginMin: Double
        let source: String
        let windowDays: Int

        enum CodingKeys: String, CodingKey {
            case minAbsD = "min_abs_d", minAbsT = "min_abs_t", minGroup = "min_group", pairs,
                 partialTodayOutcomes = "partial_today_outcomes", shortSleepMarginMin = "short_sleep_margin_min",
                 source, windowDays = "window_days"
        }
    }

    struct Methodology: Decodable, Sendable {
        struct Section: Decodable, Sendable, Hashable {
            let heading: String
            let body: String
        }

        let title: String
        let sections: [Section]
    }

    // MARK: Loading

    /// The bundled config. Missing or malformed resources are a build error, so this traps loudly in development.
    static let shared: InsightsConfig = {
        guard let config = load() else { fatalError("Insights/Resources/insights_config.json is missing or invalid") }
        return config
    }()

    static var bundledURL: URL? { Bundle.main.url(forResource: "insights_config", withExtension: "json") }

    static func load(from url: URL? = bundledURL) -> InsightsConfig? {
        guard let url, let data = try? Data(contentsOf: url) else { return nil }
        return decode(data)
    }

    static func decode(_ data: Data) -> InsightsConfig? {
        try? JSONDecoder().decode(InsightsConfig.self, from: data)
    }

    func metric(_ id: String) -> Metric? { metrics[id] }

    func marker(_ id: String) -> HealthAge.Marker? { healthAge.markers.first { $0.id == id } }

    func disclaimer(_ key: String) -> String { disclaimers[key] ?? disclaimers["general"] ?? "" }
}
